aten/src/ATen/core/NestedIntSymNodeImpl.h

```diff
@@ -183,4 +184,28 @@ class TORCH_API NestedIntSymNodeImpl : public SymNodeImpl {
   int64_t coeff_;
 };
 
+class TORCH_API UnionFind {
+    std::unordered_map<int64_t, std::pair<int64_t, size_t>> data;
+public:
+    int64_t find(int64_t x) {
+        if(data.count(x) == 0) data[x] = {x, 1};
+        if (data[x].first != x) {
+            data[x].first = find(data[x].first);
+        }
+        return data[x].first;
+    }
+    void merge(int64_t x, int64_t y) {
+        if(data.count(x) == 0) data[x] = {x, 1};
+        if(data.count(y) == 0) data[y] = {y, 1};
+        int64_t x_root = find(x), y_root = find(y);
+        if (x_root == y_root)
+            return;
+        if (data[x_root].second < data[y_root].second) {
+            std::swap(x_root, y_root);
+        }
+        data[y_root].first = x_root;
+        data[x_root].second += data[y_root].second;
+    }
```

Comment:
nit: don't put the function implementations in the header, put them in a cpp file